docs/sql-data-sources-parquet.md

```diff
@@ -252,6 +252,71 @@ REFRESH TABLE my_table;
+
+sc.hadoopConfiguration.set("parquet.encryption.kms.client.class" ,
+                           "org.apache.parquet.crypto.keytools.mocks.InMemoryKMS")
+
+// Explicit master keys (base64 encoded) - required only for mock InMemoryKMS
+sc.hadoopConfiguration.set("parquet.encryption.key.list" ,
+                   "keyA:AAECAwQFBgcICQoLDA0ODw== ,  keyB:AAECAAECAAECAAECAAECAA==")
+
+// Activate Parquet encryption, driven by Hadoop properties
+sc.hadoopConfiguration.set("parquet.crypto.factory.class" ,
+                   "org.apache.parquet.crypto.keytools.PropertiesDrivenCryptoFactory")
+
+// Write encrypted dataframe files. 
+// Column "square" will be protected with master key "keyA".
+// Parquet file footers will be protected with master key "keyB"
+squaresDF.write.
+   option("parquet.encryption.column.keys" , "keyA:square").
+   option("parquet.encryption.footer.key" , "keyB").
+parquet("/path/to/table.parquet.encrypted")
+
+// Read encrypted dataframe files
+val df2 = spark.read.parquet("/path/to/table.parquet.encrypted")
+
+{% endhighlight %}
+
+</div>
+
+
+#### KMS Client
+
+The InMemoryKMS class is provided only for illustration and simple demonstration of Parquet encryption functionality. **It should not be used in a real deployment**. The master encryption keys must be kept and managed in a production-grade KMS system, deployed in user's organization. Rollout of Spark with Parquet encryption requires implementation of a client class for the KMS server. Parquet provides a plug-in [interface](https://github.com/apache/parquet-mr/blob/apache-parquet-1.12.0/parquet-hadoop/src/main/java/org/apache/parquet/crypto/keytools/KmsClient.java) for development of such classes,
+
+<div data-lang="java"  markdown="1">
+{% highlight java %}
+
+public interface KmsClient {
+  // Wraps a key - encrypts it with the master key.
+  public String wrapKey(byte[] keyBytes, String masterKeyIdentifier);
+
+  // Decrypts (unwraps) a key with the master key. 
+  public byte[] unwrapKey(String wrappedKey, String masterKeyIdentifier);
+
+  // Use of initialization parameters is optional.
+  public void initialize(Configuration configuration, String kmsInstanceID, 
+                         String kmsInstanceURL, String accessToken);
+}
+
+{% endhighlight %}
+
+</div>
+
+An [example](https://github.com/apache/parquet-mr/blob/apache-parquet-1.12.0/parquet-hadoop/src/test/java/org/apache/parquet/crypto/keytools/samples/VaultClient.java) of such class for an open source [KMS](https://www.vaultproject.io/api/secret/transit) can be found in the parquet-mr repository. The production KMS client should be designed in cooperation with organization's security administrators, and built by developers with an experience in access control management. Once such class is created, it can be passed to applications via the `parquet.encryption.kms.client.class` parameter and leveraged by general Spark users as shown in the encrypted dataframe write/read sample above.
```

Comment:
For the hyperlinks, how about linking to 'latest' docs in Parquet?

Reply:
This doesn't quite belong in Parquet documentation.. Hashicorp Vault is a separate technology, used here as an example of an open source KMS Server. Parquet repo keeps a sample Client class for this Server, but this class is not a part of Parquet API. Moreover, Parquet unfortunately doesn't maintain a documentation of its APIs and usage, the info in https://parquet.apache.org/documentation/latest/ is very high level, and grossly outdated..

Reply:
I just mean using a URL with /latest/ in it. What do you mean this doesn't belong in the docs, if you add it here?

Reply:
@srowen will one of these two options be ok for you? (1. removing the URLs, keeping the names of the server/client; or 2. keeping the URLs to the latest versions of the server/client)

Reply:
I see, there isn't any published API doc. OK, linking to source in master is probably the closest that's possible.

Reply:
SGTM